File: lab4-parser-generator/src/test/java/lab2/LexicalAnalyzerTest.java
package lab2;

import booleanExpressions.LexicalAnalyzer;
import exceptions.LexicalAnalyzerException;
import org.antlr.v4.runtime.misc.Pair;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.*;
import java.util.stream.Stream;

class LexicalAnalyzerTest {
        private static Stream<Arguments> sequences() {
            return Stream.of(
                    Arguments.of(
                            "A", List.of(
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.VAR, "A"),
                                    LexicalAnalyzer.EOF
                            )
                    ),
                    Arguments.of(
                            "A or B", List.of(
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.VAR, "A"),
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.OR, "or"),
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.VAR, "B"),
                                    LexicalAnalyzer.EOF
                            )
                    ),
                    Arguments.of(
                            "A or B and", List.of(
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.VAR, "A"),
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.OR, "or"),
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.VAR, "B"),
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.AND, "and"),
                                    LexicalAnalyzer.EOF
                            )
                    ),
                    Arguments.of(
                            "Aor\nBand", List.of(
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.VAR, "A"),
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.OR, "or"),
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.VAR, "B"),
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.AND, "and"),
                                    LexicalAnalyzer.EOF
                            )
                    ),
                    Arguments.of(
                            "AorBand",
                            List.of(
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.VAR, "A"),
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.OR, "or"),
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.VAR, "B"),
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.AND, "and"),
                                    LexicalAnalyzer.EOF
                            )
                    ),
                    Arguments.of(
                            "and", List.of(
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.AND, "and"),
                                    LexicalAnalyzer.EOF
                            )
                    ),
                    Arguments.of(
                            "(A or B) and C", List.of(
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.LPARENT, "("),
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.VAR, "A"),
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.OR, "or"),
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.VAR, "B"),
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.RPARENT, ")"),
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.AND, "and"),
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.VAR, "C"),
                                    LexicalAnalyzer.EOF
                            )
                    ),
                    Arguments.of(
                            "A AND B", List.of(
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.VAR, "A"),
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.VAR, "A"),
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.VAR, "N"),
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.VAR, "D"),
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.VAR, "B"),
                                    LexicalAnalyzer.EOF
                            )
                    ),
                    Arguments.of(
                            "andandand", List.of(
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.AND, "and"),
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.AND, "and"),
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.AND, "and"),
                                    LexicalAnalyzer.EOF
                            )
                    ),
                    Arguments.of(
                            "   andorand", List.of(
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.AND, "and"),
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.OR, "or"),
                                    new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.AND, "and"),
                                    LexicalAnalyzer.EOF
                            )
                    ),
                    Arguments.of("", List.of(LexicalAnalyzer.EOF))
            );
        }

        private static Stream<Arguments> sequencesAsserts() {
            return Stream.of(Arguments.of("1"),
            Arguments.of("A.F")
        );
        }

        private static final Random random = new Random();

        private static Pair<String, List<LexicalAnalyzer.Token>> generateRandomTest(int len, String separator) {
            Map<String, LexicalAnalyzer.Token> tokens = Map.of(
                "xor" , new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.XOR, "xor"),
                "not" , new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.NOT, "not"),
                "in" , new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.IN, "in"),
                "and" , new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.AND, "and"),
                "or" , new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.OR, "or"),
                "C" , new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.VAR, "C"),
                "A" , new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.VAR, "A"),
                "F" , new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.VAR, "F"),
                "(" , new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.LPARENT, "("),
                ")" , new LexicalAnalyzer.Token(LexicalAnalyzer.TypeToken.RPARENT, ")")
            );
            List<LexicalAnalyzer.Token> expected = new ArrayList<>();
            List<String> test = new ArrayList<>();
            List<Map.Entry<String, LexicalAnalyzer.Token>> entries = tokens.entrySet().stream().toList();

            for (int i = 0; i < len; i++) {
                int randomInd = random.nextInt(10);
                Map.Entry<String, LexicalAnalyzer.Token> randomEntry = entries.get(randomInd);

                test.add(randomEntry.getKey());
                expected.add(randomEntry.getValue());
            }
            expected.add(LexicalAnalyzer.EOF);

            return new Pair<>(String.join(separator, test), expected);
        }

    @ParameterizedTest
    @MethodSource("sequences")
    public void test(String string, List<LexicalAnalyzer.Token> expected) throws LexicalAnalyzerException {
        LexicalAnalyzer analyzer = new LexicalAnalyzer(string);

        for (LexicalAnalyzer.Token expectedToken : expected) {
            Assertions.assertEquals(expectedToken, analyzer.getCurToken());
            analyzer.nextToken();
        }
        Assertions.assertEquals(analyzer.getCurToken(), LexicalAnalyzer.EOF);
    }

    @Test
    public void randomTest() throws LexicalAnalyzerException {
        for (int i = 0; i < 1000; i++) {
            Pair<String, List<LexicalAnalyzer.Token>> arguments = generateRandomTest(random.nextInt(100), "   ");

            test(arguments.a, arguments.b);
        }
    }

    @ParameterizedTest
    @MethodSource("sequencesAsserts")
    public void failTest(String string) {
        try {
            LexicalAnalyzer analyzer = new LexicalAnalyzer(string);
            do {
                analyzer.nextToken();
            } while (analyzer.getCurToken() != LexicalAnalyzer.EOF);
        } catch (LexicalAnalyzerException ignored) {
        }
    }

}

File: lab2-boolean-expression/src/test/kotlin/LexicalAnalyzerTest.kt
import exceptions.LexicalAnalyzerException
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.MethodSource
import java.text.ParseException
import kotlin.random.Random

class LexicalAnalyzerTest {
    companion object {
        @JvmStatic
        fun sequences() = listOf(
            Arguments.of(
                "A", listOf(
                    Token(TypeToken.VARIABLE, "A"),
                    Token(TypeToken.END)
                )
            ),
            Arguments.of(
                "A or B", listOf(
                    Token(TypeToken.VARIABLE, "A"),
                    Token(TypeToken.OR),
                    Token(TypeToken.VARIABLE, "B"),
                    Token(TypeToken.END)
                )
            ),
            Arguments.of(
                "A or B and", listOf(
                    Token(TypeToken.VARIABLE, "A"),
                    Token(TypeToken.OR),
                    Token(TypeToken.VARIABLE, "B"),
                    Token(TypeToken.AND),
                    Token(TypeToken.END)
                )
            ),
            Arguments.of(
                "Aor\nBand", listOf(
                    Token(TypeToken.VARIABLE, "A"),
                    Token(TypeToken.OR),
                    Token(TypeToken.VARIABLE, "B"),
                    Token(TypeToken.AND),
                    Token(TypeToken.END)
                )
            ),
            Arguments.of(
                "aorband",
                listOf(
                    Token(TypeToken.VARIABLE, "a"),
                    Token(TypeToken.OR),
                    Token(TypeToken.VARIABLE, "b"),
                    Token(TypeToken.AND),
                    Token(TypeToken.END)
                )
            ),
            Arguments.of(
                "and", listOf(
                    Token(TypeToken.AND),
                    Token(TypeToken.END)
                )
            ),
            Arguments.of(
                "(a or b) and c", listOf(
                    Token(TypeToken.OPEN_BRACKET),
                    Token(TypeToken.VARIABLE, "a"),
                    Token(TypeToken.OR), Token(TypeToken.VARIABLE, "b"),
                    Token(TypeToken.CLOSE_BRACKET),
                    Token(TypeToken.AND),
                    Token(TypeToken.VARIABLE, "c"),
                    Token(TypeToken.END)
                )
            ),
            Arguments.of(
                "a AND b", listOf(
                    Token(TypeToken.VARIABLE, "a"),
                    Token(TypeToken.VARIABLE, "A"),
                    Token(TypeToken.VARIABLE, "N"),
                    Token(TypeToken.VARIABLE, "D"),
                    Token(TypeToken.VARIABLE, "b"),
                    Token(TypeToken.END)
                )
            ),
            Arguments.of(
                "andandand", listOf(
                    Token(TypeToken.AND),
                    Token(TypeToken.AND),
                    Token(TypeToken.AND),
                    Token(TypeToken.END)
                )
            ),
            Arguments.of(
                "   andorand", listOf(
                    Token(TypeToken.AND),
                    Token(TypeToken.OR),
                    Token(TypeToken.AND),
                    Token(TypeToken.END)
                )
            ),
            Arguments.of("", listOf(Token(TypeToken.END))),
        )

        @JvmStatic
        fun sequencesAsserts() = listOf(
            Arguments.of("1"),
            Arguments.of("A.F")
        )

        fun generateRandomTest(len: Int, separator: String): Pair<String, List<Token>> {
            val tokens = mapOf(
                "xor" to Token(TypeToken.XOR),
                "not" to Token(TypeToken.NOT),
                "in" to Token(TypeToken.IN),
                "and" to Token(TypeToken.AND),
                "or" to Token(TypeToken.OR),
                "a" to Token(TypeToken.VARIABLE, "a"),
                "b" to Token(TypeToken.VARIABLE, "b"),
                "c" to Token(TypeToken.VARIABLE, "c"),
                "x" to Token(TypeToken.VARIABLE, "x"),
                "y" to Token(TypeToken.VARIABLE, "y"),
                "A" to Token(TypeToken.VARIABLE, "A"),
                "f" to Token(TypeToken.VARIABLE, "f"),
                "(" to Token(TypeToken.OPEN_BRACKET),
                ")" to Token(TypeToken.CLOSE_BRACKET)
            )
            val expected = mutableListOf<Token>()
            val test = mutableListOf<String>()
            val entries = tokens.entries

            for (i in 0..len) {
                val randomEntry = entries.random()

                test.add(randomEntry.key)
                expected.add(randomEntry.value)
            }
            expected.add(Token(TypeToken.END))

            return test.joinToString(separator) to expected
        }
    }

    @ParameterizedTest
    @MethodSource("sequences")
    fun test(string: String, expected: List<Token>) {
        val analyzer = LexicalAnalyzer(string.byteInputStream())

        for (expectedToken in expected) {
            analyzer.nextToken()
            Assertions.assertEquals(expectedToken, analyzer.curToken)
        }
        Assertions.assertEquals(analyzer.curToken, Token(TypeToken.END))
    }

    @Test
    fun randomTest() {
        for (i in 0..1000) {
            val arguments = generateRandomTest(Random.nextInt(100), "   ")

            test(arguments.first, arguments.second)
        }
    }

    @ParameterizedTest
    @MethodSource("sequencesAsserts")
    fun failTest(string: String) {
        val analyzer = LexicalAnalyzer(string.byteInputStream())

        Assertions.assertThrows(LexicalAnalyzerException::class.java) {
            do {
                analyzer.nextToken()
            } while (analyzer.curToken != Token(TypeToken.END))
        }
    }

}